File: python/population_genomes.py
from bisect import bisect_left
from collections import deque
from random import random

import numpy as np

from sex import Sex
from recomb_genome import RecombGenome, Diploid, CHROMOSOME_ORDER

def _pick_chroms_for_diploid(genome, recombinator):
    """
    Takes a genome and returns a diploid chromosome that is the result
    of recombination events and randomly picking a diploid for each
    autosome.
    """
    recomb_genome = recombinator.recombination(genome)
    mother = recomb_genome.mother
    father = recomb_genome.father
    starts = []
    founder = []
    offsets = recombinator._chrom_start_offset
    chrom_stop = 0
    for chrom_name in CHROMOSOME_ORDER:
        if random() < 0.5:
            tmp_diploid = mother
        else:
            tmp_diploid = father

        chrom_start = bisect_left(tmp_diploid.starts, offsets[chrom_name])
        if chrom_name != CHROMOSOME_ORDER[-1]:
            chrom_stop = bisect_left(tmp_diploid.starts,
                                     offsets[chrom_name + 1],
                                     chrom_start)
        else:
            chrom_stop = len(tmp_diploid.starts)
        starts.extend(tmp_diploid.starts[chrom_start:chrom_stop])
        founder.extend(tmp_diploid.founder[chrom_start:chrom_stop])
    return Diploid(np.array(starts, dtype = np.uint32),
                   mother.end,
                   np.array(founder, dtype = np.uint32))


def mate(mother, father, mother_recombinator, father_recombinator):
    """
    Takes a mother and father, and returns a genome for a child
    """
    assert mother is not None
    assert father is not None
    from_mother = _pick_chroms_for_diploid(mother, mother_recombinator)
    from_father = _pick_chroms_for_diploid(father, father_recombinator)
    return RecombGenome(from_mother, from_father)

def generate_genomes_ancestors(root_nodes, generator, recombinators):
    queue = deque(root_nodes)
    visited = set()
    while len(queue) > 0:
        person = queue.popleft()
        if person in visited:
            continue
        if person.mother is not None:
            assert person.father is not None
            if person.mother not in visited or person.father not in visited:
                continue
            person.genome =  mate(person.mother.genome, person.father.genome,
                                  recombinators[Sex.Female],
                                  recombinators[Sex.Male])
        else:
            person.genome = generator.generate()
        queue.extend(person.children)
        visited.add(person)

def generate_genomes(population, generator, recombinators, keep_last = None,
                     true_genealogy = True):
    assert keep_last is None or keep_last > 0
    for generation_num, generation in enumerate(population.generations):
        for person in generation.members:
            if person.genome is not None:
                continue
            if person.twin is not None and person.twin.genome is not None:
                person.genome = person.twin.genome
                continue
            if true_genealogy:
                mother = person.mother
                father = person.father
            else:
                mother = person.suspected_mother
                father = person.suspected_father
            
            if mother is None and father is None:
                person.genome = generator.generate()
                continue
            if mother is None:
                mother_genome = generator.generate()
            else:
                mother_genome = mother.genome
            if father is None:
                father_genome = generator.generate()
            else:
                father_genome = father.genome
                
            assert mother_genome is not None
            assert father_genome is not None
            person.genome = mate(mother_genome, father_genome,
                                 recombinators[Sex.Female],
                                 recombinators[Sex.Male])
        if keep_last is not None and keep_last <= generation_num:
            to_delete = population.generations[generation_num - keep_last]
            for person in to_delete.members:
                person.genome = None

File: rust/src/genome/population_genomes.rs
use std::rc::Rc;

use rand::{Rng, thread_rng};
// use rand::distributions::{IndependentSample, Range};

use population::node::Population;
use genome::genome::{Genome, CHROMOSOMES, RecombGenomeGenerator};
use genome::recombinator::{Recombinator, RecombinatorPair};
use genome::diploid::Diploid;

pub fn generate_genomes(population: &Population,
                        generator: &RecombGenomeGenerator,
                        recombinators: &RecombinatorPair,
                        true_genealogy: bool,
                        keep_last: u32) {
    let generations = &population.generations;
    for (generation_num, generation) in generations.iter().enumerate() {
        for node_ref in &generation.members {
            let mut node = node_ref.borrow_mut();
            if node.genome.is_some() {
                continue;
            }
            let mother_option;
            let father_option;
            if true_genealogy {
                mother_option = node.mother.upgrade();
                father_option = node.father.upgrade();
            } else {
                mother_option = node.suspected_mother.upgrade();
                father_option = node.suspected_father.upgrade();
            }
            if mother_option.is_none() && father_option.is_none() {
                node.genome = Some(Rc::new(generator.generate()));
                continue;
            }
            if let Some(twin) = node.twin.upgrade() {
                let twin_borrow = twin.borrow();
                let twin_mother_option;
                let twin_father_option;
                if true_genealogy {
                    twin_mother_option = twin_borrow.mother.upgrade();
                    twin_father_option = twin_borrow.father.upgrade();
                } else {
                    twin_mother_option = twin_borrow.suspected_mother.upgrade();
                    twin_father_option = twin_borrow.suspected_father.upgrade();
                }
                if twin_borrow.genome.is_some() && mother_option == twin_mother_option && father_option == twin_father_option {
                    node.genome = twin_borrow.genome.clone();
                    continue;
                }
            }
            let mother_genome;
            if mother_option.is_none() {
                mother_genome = Rc::new(generator.generate());
            } else {
                let mother = mother_option.unwrap();
                let mother_ref = mother.borrow();
                mother_genome = mother_ref.genome.as_ref().unwrap().clone();
            }
            let father_genome;
            if father_option.is_none() {
                father_genome = Rc::new(generator.generate());
            } else {
                let father = father_option.unwrap();
                let father_ref = father.borrow();
                father_genome = father_ref.genome.as_ref().unwrap().clone();
            }
            node.genome = Some(Rc::new(mate(mother_genome.as_ref(),
                                            father_genome.as_ref(),
                                            &recombinators)));
        }
        if keep_last <= generation_num as u32 {
            let to_clean_gen_num = generation_num - keep_last as usize;
            let to_clean = &population.generations[to_clean_gen_num];
            for node_ref in &to_clean.members {
                let mut node = node_ref.borrow_mut();
                node.genome = None;
            }
        }
    }
    
}

pub fn clean_genomes(population: &Population) {
    for node_ref in &population.members {
        let mut node = node_ref.borrow_mut();
        node.genome = None;
    }
}

fn pick_chroms_for_diploid(genome: &Genome, recombinator: &Recombinator)
                           -> Diploid {
    let recomb_genome = recombinator.recombination(genome);
    let mut starts = Vec::new();
    let mut founder = Vec::new();
    let mut rng = thread_rng();
    for &chrom in CHROMOSOMES {
        let tmp_diploid;
        if rng.gen() {
            tmp_diploid = &recomb_genome.mother;
        } else {
            tmp_diploid = &recomb_genome.father;
        }
        let start = recombinator.chromosomes[&chrom].chrom_start_offset;
        let start_i = match tmp_diploid.starts.binary_search(&start) {
            Ok(i) => i,
            Err(i) => i
        };
        let stop_i;
        if chrom != CHROMOSOMES[CHROMOSOMES.len() - 1] {
            let stop = recombinator.chromosomes[&(chrom + 1)].chrom_start_offset;
            stop_i = match tmp_diploid.starts.binary_search(&stop) {
                Ok(i) => i,
                Err(i) => i
            };
        } else {
            stop_i = tmp_diploid.starts.len();
        }
        starts.extend_from_slice(&tmp_diploid.starts[start_i..stop_i]);
        founder.extend_from_slice(&tmp_diploid.founder[start_i..stop_i]);
    }
    Diploid {starts: starts, founder: founder, end: recomb_genome.mother.end}
}

fn mate(mother: &Genome, father: &Genome,
        recombinators: &RecombinatorPair) -> Genome {
    // mother_recombinator: &Recombinator,
    // father_recombinator: &Recombinator) -> Genome {
    let from_mother = pick_chroms_for_diploid(mother, &recombinators.female);
    let from_father = pick_chroms_for_diploid(father, &recombinators.male);
    Genome {mother: from_mother, father: from_father}
}
